index.d.ts

```diff
@@ -0,0 +1,70 @@
+import { ComponentClass, Component, StatelessComponent, ReactNode } from 'react';
+import { Store, Dispatch, ActionCreator } from 'redux';
+
+interface ComponentDecorator<TOriginalProps, TOwnProps> {
+  (component: ComponentClass<TOriginalProps>|StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+}
+
+/**
+ * Following 3 functions cover all possible ways connect could be invoked
+ *
+ * - State: Redux state interface (the same one used by Store<S>)
+ * - TStateProps: Result of MapStateToProps
+ * - TDispatchProps: Result of MapDispatchToProps
+ * - TOwnProps: Props passed to the wrapping component
+ */
+export function connect<State, TOwnProps>(): ComponentDecorator<{ dispatch: Dispatch<State> }, TOwnProps>;
+
+export function connect<State, TOwnProps, TStateProps>(
+  mapStateToProps: MapStateToProps<State, TStateProps, TOwnProps>,
+): ComponentDecorator<TStateProps & { dispatch: Dispatch<State> }, TOwnProps>;
```

Comment:
Ok. I felt that not exposing state props and dispatch prop to parent component was actually a feature.
Is there a valid use case where we should let users overwrite HOC's props?
Or is it more that you feel type checking should try to stick to reality as much as possible?